cylc/uiserver/schema.py

```diff
@@ -396,78 +415,299 @@ def run_task_query(conn, workflow):
+            * None = no execution attempt
+        time_run:
+            * True if the job would have started running
+
+    Examples:
+        >>> _status_to_state('running')
+        (0, None, True)
+        >>> _status_to_state('waiting')
+        (None, None, None)
+
+        >>> _state_to_status(*_status_to_state('succeeded'))
+        'succeeded'
+
+    """
+    return _JOB_STATUS_TO_STATE.get(status, (None, None, None))
+
+
+def _state_to_status(
+    submit_status: Optional[int],
+    run_status: Optional[int],
+    time_run: Optional[str],
+) -> str:
+    """Derive job status from state attributes.
+
+    Args:
+        submit_status: Exit code from job submission.
+        run_status: Exit code from job execution.
+        time_run: The time the job reported that it started running.
+            (note this is interpreted as a bool, the exact value is ignored).
+
+    Returns:
+        status: The job status, e.g "submitted" or "running".
+
+    Examples:
+        >>> _state_to_status(0, None, True)
+        'running'
+        >>> _state_to_status(None, None, None)
+        'waiting'
+
+        >>> _status_to_state(_state_to_status(0, 1, True))
+        (0, 1, True)
+
+    """
+    if run_status is not None:
+        if run_status == 0:
+            status = TASK_STATUS_SUCCEEDED
+        else:
+            status = TASK_STATUS_FAILED
+    elif time_run is not None:
+        status = TASK_STATUS_RUNNING
+    elif submit_status is not None:
+        if submit_status == 0:
+            status = TASK_STATUS_SUBMITTED
+        else:
+            status = TASK_STATUS_SUBMIT_FAILED
+    else:
+        status = TASK_STATUS_WAITING
 
-    # TODO: support all arguments including states
-    # https://github.com/cylc/cylc-uiserver/issues/440
+    return status
+
+
+def run_jobs_query(
+    conn,
+    workflow,
+    ids=None,
+    exids=None,
+    states=None,
+    exstates=None,
+    tasks=None,
+):
+    """Query jobs from the database.
+
+    Args:
+        conn: Database connection.
+        workflow: Workflow ID.
+        kwargs: GraphQL sort/filter args as per cylc-flow interfaces.
+
+    """
+    # TODO: support all arguments:
+    # * [x] ids
+    # * [ ] sort
+    # * [x] exids
+    # * [x] states
+    # * [x] exstates
+    # See https://github.com/cylc/cylc-uiserver/issues/440
     jobs = []
+    where_stmts = []
+    where_args = []
+
+    # filter by cycle/task/job ID
+    if ids:
+        items = []
+        for id_ in ids:
+            item = []
+            for token, column in (
+                ('cycle', 'cycle'),
+                ('task', 'name'),
+                ('job', 'submit_num'),
+            ):
+                value = id_[token]
+                if value:
+                    if token == 'job':
+                        value = int(value)
+                    item.append(rf'{column} GLOB ?')
+                    where_args.append(value)
+            items.append(r'(' + ' AND '.join(item) + r')')
```

Comment:
F-string might be tidier?

```suggestion
        id_items.append(rf'({" AND ".join(item)})')
```